[Yerin512] 3주차 과제 To-do 리스트 테스트 작성하기

yerin512 opened the pull request:
아직 Input.jsx만 테스트 코드를 작성 중이고, 오류도 난 상태지만 기록이라도 남기고자 push했습니다...
Click 확인하는 것은 아직이고 change 이벤트 확인하는 중인데
 expect(handleChange).toBeCalledWith(taskTitle); 에서 오류가 나서 파악중입니다.
테스트 코드 작성은 처음이라 너무 어렵네요....

Q.  강의에서 나중에 따로 확인해보라한 describe - contest - it / given 을 써도 되나요?

yerin512 committed:
JSX props는 함수를 사용하면 안 되는 eslilnt 오류를 고쳐라

Why you shouldn't use inline arrow functions in JSX props.
해당 오류가 떠서 함수 선언문을 화살표 함수로 바꿔주었습니다.

yerin512 committed:
Input.jsx 를 테스트하는 코드를 작성하라

Input.jsx를 테스트하는 유닛 테스트 코드를 작성 중입니다.

hannut91 commented:
네 쓰셔도 됩니다 지금은 먼저 쉬운 것 부터 해봅시다 ㅎㅎ

yerin512 committed:
List.jsx를 테스트하기 위한 테스트 코드를 작성했습니다

할 일이 있는 경우와 없는 경우로 나눠 작성했습니다.

yerin512 committed:
Input.jsx를 테스트하기 위한 테스트 코드를 작성했습니다 (오류 수정중

yerin512 commented:
종혁님의 코드리뷰를 보고 BDD를 이용하여 + 구글링과 공식문서와 다른 분들의 코드를 보고 List.jsx는 pass가 떴으나
Input.test.jsx는 오류는 해결하지 못하고 우선 push하였습니다. 
막히긴 막히는데 무엇을 질문해야할지 잘 모르겠습니다...
혹시 테스트코드 과제 진행할 때 공식문서 보는 거 외에 다른 팁이 있을까요?

hannut91 commented:
어떤 시도를 했고 어떻게 막혔는지 알려주시면 좋을 것 같아요.

테스트는 일종의 패턴처럼 작성하게 되어서 강의에서 나오는 내용을 그대로 외워서 하시는게 좋습니다

hannut91 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,36 @@
+    expect(container).toHaveTextContent('뭐라도 하기');
+  });
+
+  context('추가 버튼을 누르기', () => {
+    const addeButton = getByText('추가');
+
+    expect(handleClick).not.toBeCalled();
+
+    fireEvent.click(addeButton);
+
+    expect(handleClick).toHaveBeenCalledTimes(1);
+  });
```
Describe-Context-It에서 Describe는 테스트의 대상을, Context는 주어진 상황이 달라질 때를 명시할 때 It는 그래서 어떻게 결과가 나오는지를 명시합니다.

```js
describe('추가 버튼 클릭', () => {
  it('onClick 핸들러를 호출한다', () => {
    const { container, getByText } = render((
        <Input
           value=""
           onChange={handleChange}
           onClick={handleClick}
        />
    ));
  
    fireEvent.click(getByText('추가'));

    expect(handleClick).toHaveBeenCalled(1); 
  });
});
```

yerin512 replied:
네! Input이라는 컴포넌트 자체를 테스트한다는 의미로 describe('Input') 이라고 크게 묶고 그 아래로 it들을 명시했습니다. 
제 생각엔 그렇게 해도 될 거같은데, 될까요?!

hannut91 reviewed (commented):
다음 문서를 읽어보세요

https://ko.javascript.info/testing-mocha

yerin512 committed:
Input.jsx 테스트 코드를 추가했습니다 (오류)

yerin512 committed:
Input.jsx의 테스트 코드를 수정 중입니다

yerin512 commented on src/Input.test.jsx:
```diff
@@ -3,34 +3,46 @@ import { fireEvent, render } from '@testing-library/react';
+      fireEvent.click(addeButton);
 
-    fireEvent.click(addeButton);
+      expect(handleClick).toBeCalled();
```
![image](https://user-images.githubusercontent.com/73337811/175316061-b31ac9ca-9a3b-4c72-b374-080a2df37e1f.png)

처음엔 이 `expect(handleClick)~` 에서 이미지와 같은 에러가 떴습니다.
error : `expect(jest.fn()).toHaveBeenCalledTimes(expected)`

하지만 구글링을 해도 왜 이 오류가 뜨는지 모르겠어서
우선 render()부분은 매번 it안에 넣어주어야한다는 말을 보고
변경해주니 해당 오류는 더 이상 뜨지않았습니다...
이게 render()를 매번 it안에 넣어준 거랑 어떤 상관이 있을까요?

hannut91 replied:
에러 메시지를 해석할 수 있어야 합니다. 

```
Expected number of calls: 1
Received number of called: 0
```

1번 호출되기를 기대했는데, 실제로 0을 받았다. 그래서 테스트가 통과하지 않았다 라는 뜻입니다.

그와 별개로 `render`는 테스트 대상이므로 `it`안에 넣어주어야 합니다. 사실 저도 잘 모르겠습니다. 그냥 사용법을 정확하게 써야 합니다. 테스트 대상은 `it`안에 써야 합니다. golang의 ginkgo같은 프레임워크는 `it`를 안쓰면 아예 에러가 나서 실행이 안됩니다.

hannut91 replied:
Testing Library문서를 보니 다음과 같은 내용을 찾았습니다.

> Please note that this is done automatically if the testing framework you're using supports the afterEach global and it is injected to your testing environment (like mocha, Jest, and Jasmine). If not, you will need to do manual cleanups after each test.

`it`안에 render를 하지 않고, 최상단에서 한 번만 render한다면, 첫 번째 테스트에서는 렌더된 결과물을 볼 수 있으니 올바르게 테스트가 되지만, 두 번 째 테스트부터는 렌더된게 클리어 되니 실패하겠죠. 

<img width="617" alt="image" src="https://user-images.githubusercontent.com/14071105/175445153-99a0c68a-e151-48ed-a473-62dddd88de30.png">
<img width="619" alt="image" src="https://user-images.githubusercontent.com/14071105/175445213-fc29510d-0383-4d39-8e83-4de3d8d0d818.png">
<img width="627" alt="image" src="https://user-images.githubusercontent.com/14071105/175445269-74e0a899-0f26-41f4-b4cf-c71dda0eb5cc.png">

테스트를 위에서 부터 하나씩 지우면 테스트 중에 첫 번째만 성공한다는 것을 볼 수 있죠

https://testing-library.com/docs/react-testing-library/api#cleanup

yerin512 replied:
아하, describe 안에 넣는다고 각 it에서도 받아 쓸수있다고 생각했는데 각 테스트마다 렌더대상을 넣어주어야하는 군요 !!
전역변수같은 개념으로 이해하고있었는데 잘못하고 있었습니다! 감사합니다!!!

hannut91 replied:
describe 안에 넣어서 계속 실행이 안되서 문제가 생기는 거라면 다음과 같이 해도 될 것 같긴하네요. 저도 다음에 시도를 해봐야겠습니다

```js
beforeEach(() => {
  render();
});
```

yerin512 committed:
Input.jsx의 테스트 코드 에러를 수정했습니다

yerin512 commented on src/Input.test.jsx:
```diff
@@ -3,46 +3,48 @@ import { fireEvent, render } from '@testing-library/react';
-      expect(input).toHaveTextContent(todoTitle);
-    });
+    expect(handleChange).toBeCalled();
+    expect(getByLabelText('할 일').value).toBe('뭐라도 하기');
```
![image](https://user-images.githubusercontent.com/73337811/175317777-d3baef11-63a9-4826-b664-d3af2025ce83.png)

또 한 번 코드 수정을 하고나니 이 부분에서 이미지처럼 오류가 났습니다.
Object.is() 메서드는 두 값이 같은 값인지 결정하는 거라고 알고있는데,
기존의 value는 "" 였고, 이제 '뭐라도 하기' 로 변경되길 기대하는 건데 왜 
`""`랑 `'뭐라도 하기'` 가 같은 값이라는 걸까요?

그래서 아예 render한 Input의 value =""를 아예 지우니까
정상적으로 테스트가 완료되더라구요!

hannut91 replied:
먼저 테스트를 위해 컴포넌트를 render하는 곳을 봅시다.

```
const { getByText, getByPlaceholderText } = render((
    <Input
      value=""
      onChange={handleChange}
      onClick={handleClick}
    />
));
```

value로 값을 넘겨주고 있죠. 그래서 주어진 값으로 input태그에 값을 넣어서 렌더링을 했을거에요. 이 렌더된 값은 언제 변경되나요? 바로 value의 값이 변경될 때 입니다. 그런데 이 값의 변경은 어디서 일어나나요? 이 컴포넌트에서는 알 수 없습니다. 왜냐하면 이 컴포넌트는 주어진 값을 렌더하고, 추가 버튼을 클릭했을 때, 핸들러 함수를 실행하는 일만 하기 때문입니다. 따라서 input태그의 값이 변경되는지 테스트 할 수 없고, 이 컴포넌트의 역할도 아니기 때문에 테스트하는 것은 바람직하지 못하다고 할 수 있습니다.

yerin512 replied:
![image](https://user-images.githubusercontent.com/73337811/175521288-fd42b661-3b5e-4ac9-9ede-b0714c3d9bda.png)
안녕하세요 트레이너님 ㅜㅜ!! 
답변 감사합니다.
그런데 이해가 가지 않는 부분이 있습니다!

해당 테스트에서는 추가 버튼을 클릭 했을 때, 핸들러 함수를 실행하는 일을 다룬 테스트가 아니고
무언가를 입력했을때 핸들러 함수에 대한 테스트인데
왜 `추가 버튼을 클릭했을 때, 핸들러 함수를 실행하는 일`만 하는 게 되는 건지 잘 이해가 가지 않습니다 ㅜㅜ

hannut91 replied:
음... 어떤 의미인지 잘 이해가 어렵네요. 표시해준 부분이 테스트가 실패한다는 말씀이신가요?

yerin512 commented:
안녕하세요 트레이너님!
이해가 되지 않는 오류들을 질문드려봤는데 질문의 의도가 이해가 되지 않으시거나 미흡했다면 죄송합니다 ㅜㅜ...!!!
우선 Input.test.jsx까지 완료하고 push했습니다!

yerin512 commented:
![image](https://user-images.githubusercontent.com/73337811/175524564-fbe0524d-fb37-419d-a304-40b04d8f3b86.png)

혹시 이부분도 답변해주실 수 있을까요?ㅎㅎ
너무 뭉뚱그려서 대상을 한 거라 좋지 않을까요?! 아니면 아예 잘못된 걸까요?!

yerin512 committed:
테스트용 데이터인 할일 목록 배열을  Tasks.js에 저장했습니다

fixtures 폴더에 할일 목록 배열을 담은 파일인 Tasks.js을 생성했습니다.

yerin512 committed:
Page.jsx의 테스트 코드를 작성했습니다

yerin512 commented:
안녕하세요 트레이너님 ! Page.jsx의 테스트 코드를 작성해서 push 했습니다.
미리 테스트 할 목록을 주석으로 적어둔 다음 하나씩 테스트 코드로 바꿔갔는데
원하는 표현이 있는데 생각이 안 날 때 `kyh196201`님과 `2scent` 코드를 많이 참고해서 작성했습니다. 
정말 많은 도움이 됐습니다!! 감사합니다!
이제 조금 감이 잡히는 거 같은데, 한 번 봐주시고 제가 감을 잘못 잡고 있는 거 같다면 말씀해주세요!ㅠㅠ

Q. describe안에 describe을 몇개로 나눠서 작성해도 되는지 궁금합니다 이번 Page.test.jsx 같은 경우는 
``` 
describe('<List/>', () => { ... 
 describe('"완료" 버튼을 누를 때' ...
```
이런식으로 중첩해서 사용했거든요!!
List 대상 안에 완료 버튼을 누른 경우를 대상으로 ~~일때 ~~를 바란다! ... 이런 의미로요!

hannut91 commented:
위의 `Describe-Context-It`에 대해서 답변을 드린 이유는 `Describe-Context`만 있어서 말씀드렸어요

hannut91 commented:
> describe안에 describe을 몇개로 나눠서 작성해도 되는지 궁금합니다

작성해도 문제는 없지만, 너무 많은 중첩이 생긴다면, 테스트 대상이 너무 넓거나 많아서 그런가? 라는 고민을 해보시면 좋을 것 같아요. 
컴포넌트를 더 작게 나눠야 하는 신호로도 볼 수 있습니다.

hannut91 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,49 @@
+    ));
+  }
+
+  beforeEach(() => jest.clearAllMocks());
```
테스트가 시작되기 전에 mocking된 것들을 클리어하고 있는 것 같아요. jest 설정에서 이 것을 자동으로 실행하도록 할 수 있습니다. 한 번 찾아보세요

https://jestjs.io/docs/configuration

yerin512 replied:
네 설정해주었습니다. 
package.jsonJest의 구성을 저장하려면 Jest가 "jest"설정을 찾는 방법을 알 수 있도록 최상위 수준에서 키를 사용해야 한다해서 
![image](https://user-images.githubusercontent.com/73337811/175808765-44d861b3-d1b1-45f1-a97f-1e02dc551b76.png)
이렇게 설정해주었는데 맞게 해준걸까요?!

hannut91 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,49 @@
+  it('입력창에 입력하면 handleChange가 실행되며 입력값이 보여진다.', () => {
+    const { getByLabelText } = renderInput();
+
+    expect(handleChange).not.toBeCalled();
```
`jest.clearAllMocks()`로 초기화를 해주고 있으니, 확인하지 않아도 될 것 같아요.

테스트에 예상 구문이 너무 많으면 중요한 것이 무엇이고 실제로 뭘 테스트하는지 파악하기가 어려워져요.

`expect(handleChange).not.toBeCalled()`이 것은 사실 테스트 대상이라기 보다, 내가 하고 싶은 테스트를 통과하기 위해서는 이러한 조건이 충족된다는 것을 가정해야돼 라는 느낌이죠. GOOSGBT라는 책에서는 그래서 이것을 `허용`과 `예상`으로 구분해서 사용합니다. jest에서 `allow`같은 것이 없어서 이렇게 사용하는 것을 보진 못했지만 GOOSGBT에서 예제대로 만약 테스트 코드를 작성한다면 이런식으로 작성할 수 있을 것 같아요.

    
```js
// 이러한 조건이 충족되어있어야 테스트 대상이 올바르게 동작해!
allow(handleChange).not.toBeCalled();

// when
fireEvent.change(something);

// then
expect(handleChange).toBeCalled();
```
참고로 이 책은 매우 어려워서 지금 단계에서 읽으실 책은 아닐거에요. 참고로만 알아주세요.

yerin512 replied:
아 초기화가 있는 경우면 필요없는 조건이군요! 
실제로 뭘 테스트할지 한눈에 파악할 수 있게끔 작성해야겠습니다 감사합니다!!

네 우선 책 제목을 적어두겠습니다 감사해요!!

hannut91 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,49 @@
+
+  function renderInput() {
+    return render((
+      <Input
+        onChange={handleChange}
+        onClick={handleClick}
+      />
```
테스트는 우리가 만든 코드의 사용법을 드러내는 살아있는 문서입니다. 즉 우리가 만든 코드의 예제에요. 실제로 `<Input />` 컴포넌트를 이렇게 쓰나요? `value`속성이 있죠. 지금은 `value`속성에 아무런 값을 주지 않아서 undefined고 결국 uncontrolled component로 동작하고 있어요.

See also
  - https://reactjs.org/docs/uncontrolled-components.html

yerin512 replied:
네 수정했습니다 ! 문서도 다시 한 번 읽어보겠습니다!

hannut91 commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,107 @@
+    expect(container).toHaveTextContent('To-do');
+  });
+
+  describe('<Input/>', () => {
```
Page컴포넌트에서 지금 하위 컴포넌트의 기능들을 테스트하고 있어요. 이러면 중복된 테스트가 작성될 가능성이 높고, 하위 컴포넌트가 변경되었을 때 이 컴포넌트의 테스트도 영향을 받을 가능성이 높아요

yerin512 replied:
네! 말씀해주신 대로 중복된 테스트를 제거했습니다!

yerin512 committed:
uncontrolled component화 되었던Input.jsx의 테스트 코드를 리팩토링했습니다

기존에 value 속성 없이 작성했던 코드에 value를 추가했습니다. 또 input에 입력시 기존에는 추가된 값도 보이게끔 하려했지만 hamdleChange만 실행되도록 수정했습니다

yerin512 committed:
자동 mocking clear가 되게끔 설정해주었습니다

package.json에서 clearMocks를 true로 설정해주었습니다.

yerin512 committed:
Page.jsx의 테스트 코드를 리팩토링 했습니다

하위 컴포넌트에서 끝낸 중복 테스트 코드를 없애고 Page 화면 내에서 보이는 것들만 수행하도록 수정했습니다

yerin512 committed:
App.jsx의 테스트 코드를 작성했습니다

기능에 대한 통합 테스트 코드만 작성했습니다.

yerin512 commented:
안녕하세요 트레이너님!
아직 너무 많이 부족하단 생각이 들지만...
App 컴포넌트까지 테스트 코드 작성하여 push 했습니다!
피드백 부탁드립니다 ㅜㅜ...! 늘 감사합니다
